### PR TITLE
Load parallel.cpp -f file.txt external prompt file

### PR DESCRIPTION
**Pull request by pudepiedj**
This branch includes amendments to three files in `./llama.cpp/examples` necessary to implement the external prompt file option `-f file.txt` that arises in `./bin/parallel --help`. The three affected files are:
```
common.h: add new params.prompt_file placeholder to gpt_params definition
common.cpp: assign the name of the file in the command-prompt using argv[i]
parallel.cpp: add slice function to segment params.prompt_file; add code to assign the segments to k_prompts and overwrite the default values; display the contents of the external prompts file; add a datetime stamp to the final report including the name of the external file using #include <ctime> etc.
```
Command-line code (second run example with `-ns 128`):
```
% ./build/bin/parallel -m ./models/llama-2-13b/ggml-model-q8_0.gguf -f "ParallelQuestions.txt" -n 512 -t 1 -s 3456 -ngl 100 -c 8192 -np 4 -ns 16 -cb
```
Example output from two different runs on M2 MAX 32GB and MacOS Sonoma 14.0 (omitting the initialisation):
```
llama_new_context_with_model: n_ctx      = 8192
llama_new_context_with_model: freq_base  = 10000.0
llama_new_context_with_model: freq_scale = 1
llama_new_context_with_model: kv self size  = 6400.00 MB
llama_new_context_with_model: compute buffer total size = 691.88 MB
llama_new_context_with_model: max tensor size =   166.02 MB
ggml_metal_add_buffer: allocated 'data            ' buffer, size = 13190.58 MB, (13191.20 / 21845.34)
ggml_metal_add_buffer: allocated 'kv              ' buffer, size =  6402.00 MB, (19593.20 / 21845.34)
ggml_metal_add_buffer: allocated 'alloc           ' buffer, size =   686.02 MB, (20279.22 / 21845.34)

Now printing the external prompt file ParallelQuestions.txt

 1 prompt: What do you know about Hobbits?
 2 prompt: What is quantum field theory?
 3 prompt: Why did the chicken cross the road?
 4 prompt: Who is the president of the United States?
 5 prompt: How do I run CMake on MacOS?
 6 prompt: Do you agree that C++ is a really finicky language compared with Python3?
 7 prompt: Is it a good idea to invest in technology?
 8 prompt: Do you like Wagner's Ring?
 9 prompt: Do you think this file input option is really neat?
10 prompt: What should we all do about climate change?
11 prompt: Is time-travel possible within the laws of current physics?
12 prompt: Is it like anything to be a bat?
13 prompt: Once the chicken has crossed the road, does it try to go back?
14 prompt: Who is the greatest of all musical composers?
15 prompt: What is art?
16 prompt: Is there life elsewhere in the universe?
17 prompt: What is intelligence?
18 prompt: What is the difference between knowledge and intelligence?
19 prompt: Will religion ever die?
20 prompt: Do we understand ourselves?
21 prompt: What is the best way to cook eggs?
22 prompt: If you cannot see things, on what basis do you evaluate them?
23 prompt: Explain the role of the np junction in photovoltaic cells?
24 prompt: Is professional sport a good or bad influence on human behaviour?
25 prompt: Is capital punishment immoral?
26 prompt: Should we care about other people?
27 prompt: Who are you?
28 prompt: Which sense would you surrender if you could?
29 prompt: Was Henry Ford a hero or a villain?
30 prompt: Do we need leaders?
31 prompt: What is nucleosynthesis?
32 prompt: Who is the greatest scientist of all time so far?


main: Simulating parallel requests from clients:
main: n_parallel = 4, n_sequences = 16, cont_batching = 1, system tokens = 305

main: Evaluating the system prompt ...

Processing requests ...

main: clearing the KV cache
Client   0, seq    0, started decoding ...
Client   1, seq    1, started decoding ...
Client   2, seq    2, started decoding ...
Client   3, seq    3, started decoding ...
Client   2, seq   2/ 16, prompt   15 t, response   63 t, time  9.33 s, speed  8.36 t/s, cache miss 0  
Input:    Is it a good idea to invest in technology?
Response: It depends on your personal circumstances and financial situation. Investing in technology can be a way to diversify your portfolio and potentially generate higher returns than traditional investments like stocks and bonds. However, it is important to do your research and understand the risks before making any investment decisions.

Client   2, seq    4, started decoding ...
Client   3, seq   3/ 16, prompt   14 t, response   65 t, time  9.63 s, speed  8.20 t/s, cache miss 0  
Input:    What should we all do about climate change?
Response: We should all do something about climate change. Everyone can make a difference by reducing their carbon footprint, conserving energy and water, eating less meat, and using public transportation or biking instead of driving whenever possible. We should also support policies that promote renewable energy sources like solar and wind power.

```
**Many lines omitted.** This from the end of a run with `-ns 128`.
```
Client   0, seq 121/128, prompt   21 t, response  149 t, time 55.81 s, speed  3.05 t/s, cache miss 103  
Input:    Once the chicken has crossed the road, does it try to go back?
Response: It is difficult to say for sure what the chicken's motivations are. Chickens may cross roads for a variety of reasons, such as searching for food or water, escaping predators, or simply exploring their surroundings. If a chicken has crossed a road once, it is possible that it will attempt to go back across the same road in the future if it finds something worth returning to on the other side. However, it is also possible that the chicken may not want to cross the road again, depending on its experiences and motivations at the time. Ultimately, it is hard to know for sure what a particular chicken's thoughts or intentions are when it comes to crossing roads.

main: clearing the KV cache

RUN PARAMETERS as at Sat Sep 30 17:05:27 2023

main: n_parallel = 64, n_sequences = 128, cont_batching = 1, system tokens = 305
external prompt file (if any): ParallelQuestions.txt

Total prompt tokens:   1711, speed: 14.74 t/s
Total gen tokens:      9152, speed: 78.82 t/s
Total speed (AVG):           speed: 93.55 t/s
Cache misses:           103


llama_print_timings:        load time =   647.84 ms
llama_print_timings:      sample time =  6313.92 ms /  9280 runs   (    0.68 ms per token,  1469.77 tokens per second)
llama_print_timings: prompt eval time = 106014.63 ms / 11134 tokens (    9.52 ms per token,   105.02 tokens per second)
llama_print_timings:        eval time =  2807.62 ms /    34 runs   (   82.58 ms per token,    12.11 tokens per second)
llama_print_timings:       total time = 116116.12 ms
ggml_metal_free: deallocating
```

### CODE REVIEW

File: cmake_all.sh
Comment (ggerganov) [inline cmake_all.sh:1]: Not needed

File: ParallelQuestions.txt
Comment (ggerganov) [inline ParallelQuestions.txt:1]: Move to `prompts` change name to `parallel-questions.txt`

File: common/common.h [hunk 1]
```diff
@@ -79,6 +79,7 @@ struct gpt_params {
     std::string model_draft       = "";                              // draft model for speculative decoding
     std::string model_alias       = "unknown"; // model alias
     std::string prompt            = "";
+    std::string prompt_file       = "";  // store for external prompt file name
```
Comment (ggerganov) [inline common/common.h:82]:
```suggestion
    std::string prompt_file       = "";  // store the external prompt file name
```

File: examples/parallel/parallel.cpp [hunk 2]
```diff
@@ -70,6 +72,22 @@ struct client {
     std::vector<llama_token> tokens_prev;
 };
 
+static void printDateTime() {
+    std::time_t currentTime = std::time(nullptr);
+    std::cout << "\n\033[35mRUN PARAMETERS as at \033[0m" << std::ctime(&currentTime);
```
Comment (ggerganov) [inline examples/parallel/parallel.cpp:77]: We don't use `std::cout` in this project

File: examples/parallel/parallel.cpp [hunk 3]
```diff
@@ -70,6 +72,22 @@ struct client {
...
+}
+
+// Define a split string function to ...
+static std::vector<std::string> splitString(const std::string& input, char delimiter) {
```
Comment (ggerganov) [inline examples/parallel/parallel.cpp:81]: snake_case

**Review by ggerganov — CHANGES_REQUESTED** [review]

File: examples/parallel/parallel.cpp (hunk 2 above)
Comment (pudepiedj) [reply examples/parallel/parallel.cpp:77]: Changed to printf()

File: examples/parallel/parallel.cpp (hunk 3 above)
Comment (pudepiedj) [reply examples/parallel/parallel.cpp:81]: Changed.

File: common/common.h (hunk 1 above)
Comment (pudepiedj) [reply common/common.h:82]: Changed.

File: cmake_all.sh
Comment (pudepiedj) [reply cmake_all.sh:1]: No problem but what to use instead?

**Comment by cebtenzzre** [comment]
> Is your preference for a new PR after this kind of updating?

No, please keep the existing PR. You may `git rebase -i` and force-push to your branch if you would like to clean up the history, but it doesn't really matter because the PR will be squashed into a single commit before merging.

File: cmake_all.sh
Comment (pudepiedj) [reply cmake_all.sh:1]: Deleted. All changes are on the new branch pushed a few moments ago. New branch is <origin/Update-load-parallel-prompt-file>.

File: ParallelQuestions.txt
Comment (pudepiedj) [reply ParallelQuestions.txt:1]: Done. All changes are on the new branch pushed a few moments ago. New branch is <origin/Update-load-parallel-prompt-file>.

File: examples/parallel/parallel.cpp (hunk 2 above)
Comment (pudepiedj) [reply examples/parallel/parallel.cpp:77]: All changes to `parallel.cpp` are on the new branch pushed a few moments ago. New branch is <origin/Update-load-parallel-prompt-file>. Longer external files produce longer processing times so I have tweaked llama.cpp LOG_TEE from %8.2f -> %10.2f as well at the end.

**Comment by pudepiedj** [comment]
It's interesting to use the 100 questions in `/examples/jeopardy/questions.txt` as the external prompt file to measure the difference in performance between simultaneous and sequential processing.
Without simultaneous processing `-np 1` the file takes roughly 233 seconds on an M2 MAX (38 core) and Sonoma 14.0; with `-np 64` simultaneous processes and continuous batching `-cb` it takes about 43 seconds.
The command-line (`-c 16384` is slightly too big to run with `f16`):
```
% ./build/bin/parallel -m ./models/llama-2-7b/ggml-model-f16.gguf -f "examples/jeopardy/questions.txt" -n 256 -t 1 -ngl 100 -c 8192 -s 1234 -np 64 -ns 100 -cb
```
Since memory is critical it's worth noting the resources used and how parsimonious the system allocation is with `device.recommendedMaxWorkingSetSize':
```
ggml_metal_init: hasUnifiedMemory              = true
ggml_metal_init: recommendedMaxWorkingSetSize  = 21845.34 MB
ggml_metal_init: maxTransferRate               = built-in GPU
llama_new_context_with_model: compute buffer total size = 557.88 MB
llama_new_context_with_model: max tensor size =   250.00 MB
ggml_metal_add_buffer: allocated 'data            ' buffer, size = 12853.73 MB, (12854.36 / 21845.34)
ggml_metal_add_buffer: allocated 'kv              ' buffer, size =  4098.00 MB, (16952.36 / 21845.34)
ggml_metal_add_buffer: allocated 'alloc           ' buffer, size =   552.02 MB, (17504.38 / 21845.34)
```

**Comment by pudepiedj** [comment]
> > Thank you. All changes to <origin/load-parallel-prompt-file> now pushed to <origin/Update-load-parallel-prompt-file> which I hope I have done correctly this time.
> 
> Could you please push your changes to the load-parallel-prompt-file branch so they appear here?

OK this is what I did. I hope it's right. I've looked at the files in **load-parallel-prompt-file** and they appear to have been changed correctly. Please let me know if I have done something wrong (again)!
```
(base) edsilm2@JCPM2 llama.cpp % git remote add llama.cpp https://github.com/pudepiedj/llama.cpp.git
(base) edsilm2@JCPM2 llama.cpp % git fetch llama.cpp
From https://github.com/pudepiedj/llama.cpp
 * [new branch]      Update-load-parallel-prompt-file -> llama.cpp/Update-load-parallel-prompt-file
 * [new branch]      load-parallel-prompt-file        -> llama.cpp/load-parallel-prompt-file
 * [new branch]      master                           -> llama.cpp/master
(base) edsilm2@JCPM2 llama.cpp % git push llama.cpp Update-load-parallel-prompt-file:load-parallel-prompt-file
Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
To https://github.com/pudepiedj/llama.cpp.git
   ce10861..bf8c4df  Update-load-parallel-prompt-file -> load-parallel-prompt-file
```

File: examples/parallel/parallel.cpp [hunk 4]
```diff
@@ -392,10 +395,10 @@ int main(int argc, char ** argv) {
 
     const auto t_main_end = ggml_time_us();
 
-    printDateTime();
+    print_date_time();
 
     LOG_TEE("\n%s: n_parallel = %d, n_sequences = %d, cont_batching = %d, system tokens = %d\n", __func__, n_clients, n_seq, cont_batching, n_tokens_system);
-    std::cout << "external prompt file (if any): " << params.prompt_file << "\n\n";
+    printf("external prompt file (if any): %s \n\n", params.prompt_file.c_str());
```
Comment (cebtenzzre) [inline examples/parallel/parallel.cpp:401]: Unnecessary trailing whitespace

File: cmake_all.sh
Comment (ggerganov) [inline cmake_all.sh:1]: Either rename to `cmake-all.sh` or remove this script

**Review by ggerganov — APPROVED** [review]

File: cmake_all.sh
Comment (cebtenzzre) [reply cmake_all.sh:1]: I think it should be removed.

Comment (pudepiedj) [reply cmake_all.sh:1]: It is already in my .gitignore so what else must I do to remove it?

Comment (cebtenzzre) [reply cmake_all.sh:1]: Please remove your changes to the .gitignore and remove the file from the index with `git rm --cached cmake_all.sh`.

File: examples/parallel/parallel.cpp [hunk 5]
```diff
@@ -357,13 +395,21 @@ int main(int argc, char ** argv) {
...
 
-    LOG_TEE("\n\n");
+    print_date_time();
+
+    LOG_TEE("\n%s: n_parallel = %d, n_sequences = %d, cont_batching = %d, system tokens = %d\n", __func__, n_clients, n_seq, cont_batching, n_tokens_system);
+    if (params.prompt_file.empty()) {
+        params.prompt_file = "used built-in defaults";
+    }
+    printf("External prompt file: \033[32m%s\033[0m\n", params.prompt_file.c_str());
+    printf("Model and path used: \033[32m%s\033[0m\n\n", params.model.c_str());
```
Comment (ggerganov) [inline examples/parallel/parallel.cpp:405]:
```suggestion
    LOG_TEE("External prompt file: \033[32m%s\033[0m\n", params.prompt_file.c_str());
    LOG_TEE("Model and path used:  \033[32m%s\033[0m\n\n", params.model.c_str());
```

**Comment by pudepiedj** [comment]
Thank you. I appreciate your patience!

**Pull request merged**